### PR TITLE
Skip messages from aborted transactions

### PR DESCRIPTION
**Pull request by VictorDenisov**

### CODE REVIEW

**Comment by VictorDenisov** [comment]
I realized that it needs to take into account the config option.

**Comment by VictorDenisov** [comment]
Done.

**Comment by VictorDenisov** [comment]
Sure. Newer versions of kafka allow us to use transactions for consuming and producing messages.

In order to mark the committed and aborted messages kafka brokers use control messages. Here is the original design document for transactions and onetime delivery: https://docs.google.com/document/d/11Jqy_GjUGtdXJK94XGsEIK7CP1SnQGdp2eF0wSw9ra8/edit#heading=h.scppt9di2ipj
As a result in addition to data messages we can encounter control message in a topic on which transactions were used.

Kafka clients are supposed keep those control messages hidden from the end user. Kafka clients need to parse those control messages and identify their type: abort or commit. Abort messages mean that the preceding messages should be dropped and never returned to the user. Commit messages mean that the preceding messages need to be returned to the end user.

Control messages can appear in a different batch than the messages they are related to. As a result I can't hide control messages behind the Batch interface. They should be returned from the Batch and can only be filtered out in the Reader.

Java libraries allow you to allow the driver pass through the uncommitted messages. So I figured I would allow this option in this library as well.

In conclusion this change makes Readers transactions aware.

**Comment by VictorDenisov** [comment]
@achille-roussel Is there anything else I can do to facilitate the merge? It's a pretty useful independent change. It allows go clients to successfully consume from topics that were populated by java clients with transactions.

**Comment by Jarema** [comment]
Did this topic die?

**Comment by achille-roussel** [comment]
@Jarema would you like to take over this work?

**Comment by VictorDenisov** [comment]
I can finish it if you are interested in merging @achille-roussel

**Comment by achille-roussel** [comment]
Yep! Let’s get this through the finish line!

**Comment by dotnwat** [comment]
We at https://github.com/vectorizedio/redpanda (drop in replacement for kafka) are very excited to see this support land.

One thing I wanted to check on is that non-transaction / unknown control messages are ignored. Here is the corresponding case in librdkafka where an unknown control message is ignored (https://github.com/edenhill/librdkafka/blob/master/src/rdkafka_msgset_reader.c#L862). I briefly looked at this PR and it seemed like there was an assumption that all the control messages were transactional / known, but wanted to confirm.

Thanks!

**Comment by korzha** [comment]
@achille-roussel Interested in the change. Any help needed to push it forward?

**Comment by rhasanoff** [comment]
Hello @achille-roussel @VictorDenisov 

We are using xk6-kafka reader and it uses kafka-go. We would like to skip reading the control batch messages. Are there any plans to merge this change?
